demos/spectre_v2.cc

```diff
@@ -0,0 +1,296 @@
+/*
+ * Copyright 2019 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * Implements Spectre variant 2 (branch target injection), same address space
+ * version.
+ * Compared to competing approaches we don't bother to reverse-engineer branch
+ * predictors. Since we need a universal and robust solution we have to
+ * abstract from their peculiarities, so our solution is based on a brute-force
+ * approach.
+ * We flood the BTB with a plenty of jumps that do not follow typical patterns.
+ * In this way we interleave jumps to real fetchers that fetch safe values with
+ * jumps to dummy fetchers that do not fetch anything, but they get the unsafe
+ * values. That leads to misspeculations (some of the dummy fetchers jumps are
+ * misspeculated somewhere into the real fetcher destinations).
+ **/
+
+#include <array>
+#include <cstring>
+#include <iostream>
+
+#include "cache_sidechannel.h"
+#include "instr.h"
+
+const char *public_data = "Hello, world!";
+const char *private_data = "It's a s3kr3t!!!";
+
+// Recursive templates to repeat code without using .rept and inline assembly.
+// Generates NOP instructions.
+template<unsigned int N> struct Nops {
+  SAFESIDE_ALWAYS_INLINE static inline void generate() {
+    GenerateNop();
+    Nops<N - 1>::generate();
+  }
+};
+
+template<> struct Nops<0> {
+  SAFESIDE_ALWAYS_INLINE static inline void generate() {
+    // Empty.
+  }
+};
+
+// Fetches an address N-times in a real reader.
```

Comment:
What's a real reader here? As in what's the significance of the "real" in the descriptor "real" reader? Real as opposed to what?

Reply:
As I read down further it seems like the distinction is a real reader corresponds to the real fetchers you mentioned? And later dummy readers are the dummy fetchers? If that's a correct reading, I'd update the comments here to match the terminology you used before fetcher --> reader or vice versa reader --> fetcher, so the relationship is clearer.

Reply:
Real is indeed opposed to dummy. Real readers are FirstActualRead and SecondActualRead. United the terminology.